### ISSUE TITLE
Sometimes it will appear that indexdb does not exist, and the page will always be blank

### ISSUE THREAD

**Issue by qiaogaolong** [piotr-cz/swr-idb-cache#5]
![image](https://user-images.githubusercontent.com/35677796/205796504-8460b249-f775-438c-b810-4cda382f611b.png)

**Comment by piotr-cz** [comment]
I'm sorry, but I don't understand the issue, can you rephrase it  
or provide more context, perhaps how your implementation of the `StorageHandler` interfacelooks like?

**Comment by qiaogaolong** [comment]
@piotr-cz Hello, does the cacheProvider definitely return a value?

**Comment by piotr-cz** [comment]
This behavior is expected:

https://github.com/piotr-cz/swr-idb-cache/blob/6feb06315350c6d7c0411eeb9d6a4a5db8acca51/src/use-cache-provider.ts#L9-L15

The reason is that initializing `IDBPDatabase` and in effect `cacheProvider` is asynchronous (non-blocking) and so takes *some* time.

How long it takes depends on the CPU of the device and size of cached data. For example on low budget mobile phones it may be even few seconds. On desktop this may be the matter of milliseconds if any.

That's why the [docs suggest](https://github.com/piotr-cz/swr-idb-cache#setup) to render fallback component during the initialization:

```tsx
// Not ready yet, render fallback component
if (!cacheProvider) {
  return <div>Initializing cache…</div>
}
```

If you would like to initialize cache immediately, you'd have to use synchronous (blocking) API such as Web Storage API as in SWR Docs > Cache > [LocalStorage Based Persistent Cache](https://swr.vercel.app/docs/advanced/cache#localstorage-based-persistent-cache).

However it gives you less control of handling the state when data is populated - browser UI freezes until DOM blocking operation is finished.

There's another case when `cacheProvider` value would be undefined - when an error would be thrown during it's initialization. But then you'd see it in the DevTools console. 

Refrences:
- Mozzilla Hacks > [There is no simple solution for local storage](https://hacks.mozilla.org/2012/03/there-is-no-simple-solution-for-local-storage/)

**Comment by piotr-cz** [comment]
@qiaogaolong Does my answer above answer your concerns?

**Comment by qiaogaolong** [comment]
@piotr-cz  Hello, the fallback component is presented during initialization. If you use it, the page may not come out. I wonder if there is a way to return a state to display the page after 5 seconds or a few seconds without cacheProvider

**Comment by qiaogaolong** [comment]
@piotr-cz Hello, I am currently handling it like this, no matter whether the data is obtained or not, the page will be displayed directly after two seconds
![image](https://user-images.githubusercontent.com/35677796/207534533-8f77c5fd-8789-4995-9ae4-5b4be305826d.png)
Thank you for your answer

**Comment by piotr-cz** [comment]
Are you saying that after cache has been initialized, react still renders fallback component - even that cache provider has been resolved (isn't `undefined` anymore)?

In that case can you provide minimal reproducible example, for example on [stackblitz.com](https://stackblitz.com/fork/react-ts)?

I've used this package in two production apps used by thousands of users so I'm confident to say that this issue doesn't come from the package itself, but from specific integration.

**Comment by qiaogaolong** [comment]
@piotr-cz  I can’t give an example here, because it can’t be reproduced. It’s just a problem with the client computer. Not everyone has this problem. When I remotely controlled his computer, I found that there was nothing in his indexdb, and then the page kept showing the fallback component. , I feel that indexdb has not been parsed, I don’t know if the customer has set something

**Comment by piotr-cz** [comment]
In that case I would try to see if there are any errors or warnings in the DevTools console.

Anyway strange thing is that cacheProvider is being initialized, but internal state of the react component doesn't update and you have to force it with setTimeout.

By the way, which React version are you using and if it's v18 do you use it's concurrent mode?

**Comment by qiaogaolong** [comment]
@piotr-cz Yes, but there is no error message from the console，Using v18 I'm not using concurrent mode

**Comment by qiaogaolong** [comment]
@piotr-cz Yes, but there is no error message from the console，Using v18 I used concurrent mode

**Comment by qiaogaolong** [comment]
@piotr-cz Yes, but there is no error message from the console，Using v18 I used concurrent mode
![image](https://user-images.githubusercontent.com/35677796/208005847-76581469-57ec-471d-8649-725a39b6fb1a.png)

**Comment by piotr-cz** [comment]
To be honest, I haven't tested the combination of React concurrent mode and custom storage handler.
It *should not* create problems but it I can't say for sure.

**Comment by qiaogaolong** [comment]
But now it is solved with a timer, I will close it first, thank you for your answer

**Comment by piotr-cz** [comment]
If you ever find out what's causing this, please let me know.

**Comment by yingyuk** [comment]
In Firefox private browsing mode, indexedDB is not available. This leads to the cacheProvider being undefined, causing the page content to be unable to render. The onError event will also not be executed. In this scenario, it is not possible to render the page directly without using caching.

```jsx
export default function Layout() {
  const [error, setError] = useState()
  const cacheProvider = useCacheProvider({
    dbName: 'db',
    storeName: 'store',
    onError(err) {
      // onError will not be executed in Firefox private browsing mode.
      console.log(err)
      setError(err)
    },
  })

  if (!cacheProvider && !error) {
    return <div>Loading</div>
  }

  return (
    <SWRConfig
      value={{
        provider: cacheProvider,
      }}
    >
      <div>content</div>
    </SWRConfig>
  )
}
```

**Comment by piotr-cz** [comment]
In that case, maybe the library should check for IndexedDB being available in if not execute onError?

```js
if ( !window.indexedDB) {
  onError('indexedDB is not available')
}
```

**Comment by yingyuk** [comment]
Yes, we can check if indexedDB is available. 
```js
// In Firefox's private browsing mode, !!indexedDB will return true.
if (typeof window.indexedDB === "undefined") {
  console.log("IndexedDB is not supported in this browser.");
} else {
  const request = window.indexedDB.open("testDB");

  request.onerror = function (event) {
    onError(event.target.error);
    console.log("IndexedDB error:", event.target.error);
  };

  request.onsuccess = function (event) {
    // IndexedDB is working in this browser
    const db = event.target.result;
    db.close();
  };
}
```
